Check whether a git repo is up-to-date for upgrade-interactive

mertkahyaoglu opened the pull request:
Fixes #4896

mertkahyaoglu commented:
For some reason, `upgrade-interactive` test is passing on my local and failing here because of timeout.

rally25rs commented:
First thing I noticed is that there are several linting errors. If you run `yarn test` in the project it should list them. The build process has a pretty strict set of eslint rules, and uses Flow to enforce some type restrictions.

```
~/Projects/yarn ((63c2a3a2...)) : yarn test
yarn run v1.3.2
$ yarn lint && yarn test-only
$ eslint . && flow check
Error: src/package-request.js:350
350:     const {requests: reqPatterns, workspaceLayout} = await install.fetchRequestFromCwd();
                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Promise. This type is incompatible with
682: declare function $await<T>(p: Promise<T> | T): T;
                                   ^^^^^^^^^^^^^^ union: type application of class `Promise` | type parameter `T` of await. See lib: /private/tmp/flow/flowlib_38106ab7/core.js:682
  Member 1:
  682: declare function $await<T>(p: Promise<T> | T): T;
                                     ^^^^^^^^^^ type application of class `Promise`. See lib: /private/tmp/flow/flowlib_38106ab7/core.js:682
  Error:
  425:     const isDepOld = ({current, latest, wanted, isOldHash}) =>
                                                       ^^^^^^^^^ property `isOldHash`. Property not found in
  677:     static all<Elem, T:Iterable<Elem>>(promises: T): Promise<$TupleMap<T, typeof $await>>;
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ object type. See lib: /private/tmp/flow/flowlib_38106ab7/core.js:677
  Member 2:
  350:     const {requests: reqPatterns, workspaceLayout} = await install.fetchRequestFromCwd();
                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `T` of await
  Error:
  350:     const {requests: reqPatterns, workspaceLayout} = await install.fetchRequestFromCwd();
                  ^^^^^^^^ property `requests`. Property not found in
  350:     const {requests: reqPatterns, workspaceLayout} = await install.fetchRequestFromCwd();
                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Promise

Error: src/package-request.js:390
390:             const {hash} = versionUtil.explodeHashedUrl(resolved);
                                                             ^^^^^^^^ null. This type is incompatible with the expected param type of
  3: export function explodeHashedUrl(url: string): {url: string, hash: string} {
                                           ^^^^^^ string. See: src/util/version.js:3

Error: src/package-request.js:390
390:             const {hash} = versionUtil.explodeHashedUrl(resolved);
                                                             ^^^^^^^^ undefined. This type is incompatible with the expected param type of
  3: export function explodeHashedUrl(url: string): {url: string, hash: string} {
                                           ^^^^^^ string. See: src/util/version.js:3


Found 3 errors
error Command failed with exit code 2.
```

rally25rs commented on src/package-request.js:
```diff
@@ -369,16 +372,34 @@ export default class PackageRequest {
           url = normalized.range;
+
+          // check whether the exotic version has a new commit on git.
+          if (exoticResolver === GitResolver) {
```
Flow gets mad at this because it thinks `resolved` can be null/undefined, but the function `versionUtil.explodeHashedUrl(resolved)` does not accept null.

You can fix this by adding a check to the `if`:

```
if (exoticResolver === GitResolver && resolved) {
```

rally25rs commented on src/package-request.js:
```diff
@@ -391,6 +412,7 @@ export default class PackageRequest {
           wanted,
           latest,
           url,
+          isOldHash,
```
Flow gets mad at this because, although it looks like a plain JS object is being made here, it actually wants an object that fits the `Dependency` type, which is defined in `src/types.js`. That type definition does not have a field named `isOldHash`, so to fix it, we need to add it:

_in src/types.js_ add `isOldHash` as an optional boolean field:

```
export type Dependency = {
  // ...other existing properties omitted...
  isOldHash?: boolean,
};
```

mertkahyaoglu commented:
@rally25rs I fixed the tests. I was introducing a new git module that I created for the test, `yarn-test`, and it was not being resolved. I reverted that test and used already existing one `max-safe-integer`. I also removed the assertion from `__tests__/commands/outdated.js` since it is not outdated with these changes anymore.

johnunclesam commented:
How can we help?

frederikhors commented:
Is there any news on this? How to fix the tests?

mertkahyaoglu commented:
I'm stuck on the tests. Closing this so that others can take over.
